arc_cache.go

```diff
@@ -105,6 +111,29 @@ func (b *arccache) GetSize(k cid.Cid) (int, error) {
 	return blockSize, err
 }
 
+func (b *arccache) View(k cid.Cid, callback func([]byte) error) error {
+	// shortcircuit and fall back to Get if the underlying store doesn't support
+	// Viewer.
+	if b.viewer == nil {
+		blk, err := b.Get(k)
+		if err != nil {
+			return err
+		}
+		return callback(blk.RawData())
+	}
+
+	if !k.Defined() {
+		log.Error("undefined cid in arc cache")
+		return ErrNotFound
+	}
+
+	if has, _, ok := b.hasCached(k); ok && !has {
```

Comment:
the semantics that if a viewer is set, then the calls to `View` will only run `View` on items that are in the cache is not fully intuitive.
Is there a reason not to attempt to go back to the underlying store when not cached?

Reply:
Here the cache responds that it has seen this element and it was confirmed to not exist (cache returned `ok && !has`).

Reply:
We do the same in `Get` if you look few lines lower.

Reply:
I think naming is the problem. `hasCached` naturally read as: "is this item in the cache?" instead of "check if the item is in the _has_ cache".

@willscott this just short circuits the view if we _know_ that the item doesn't exist, because we have a cache entry for it, and it's negative.